website/source/docs/platform/k8s/run.html.md

```diff
@@ -0,0 +1,234 @@
+everything is stable, begin by decreasing the `updatePartition` value by one,
+and running `helm upgrade` again. This should cause the first Consul server
+to be stopped and restarted with the new image.
+
+Wait until the Consul server cluster is healthy again (30s to a few minutes)l
+then decrease `updatePartition` and upgrade again. Continue until
+`updatePartition` is `0`. At this point, you may remove the
+`updatePartition` configuration. Your server upgrade is complete.
+
+#### Upgrading Consul Clients
+
+With the servers upgraded, it is time to upgrade the clients. To upgrade
+the clients, set the `client.image` value to the desired Consul version.
+Then, run `helm upgrade`. This will upgrade the clients in batches, waiting
+until the clients come up healthy before continuing.
+
+## Architecture
+
+We recommend running Consul on Kubernetes with the same
+[general architecture](/docs/internals/architecture.html)
+as running it anywhere else. There are some benefits Kubernetes can provide
+that eases operating a Consul cluster and we document those below. The standard
+[production deployment guide](/docs/guides/deployment.html) is still an
+important read even if running Consul within Kubernetes.
+
+Each section below will outline the different components of running Consul
+on Kubernetes and an overview of the resources that are used within the
+Kubernetes cluster.
+
+### Server Agents
+
+The server agents are run as a **StatefulSet**, using persistent volume
+claims to store the server state. This also ensures that the
+[node ID](/docs/agent/options.html#_node_id) is persisted so that servers
+can be rescheduled onto new IP addresses without causing issues. The agents
+are configured with
+[anti-affinity](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity)
+rules so that they are placed on different nodes. A readiness probe is
+configured that marks the pod as ready only when it has established a leader.
+
+A **Service** is registered to represent the servers and expose the various
+ports. The DNS address of this service is used to join the servers to each
+other without requiring any other access to the Kubernetes cluster. The
+service is configured to publish non-ready endpoints so that it can be used
+for joining during bootstrap and upgrades.
+
+Additionally, a **PodDisruptionBudget** is configured so the Consul server
+cluster maintains quorum during voluntary operational events. The maximum
+unavailable is `(n/2)-1` where `n` is the number of server agents.
+
+### Client Agents
+
+The client agents are run as a **DaemonSet**. This places one agent
+(within its own pod) on each Kubernetes node.
+The clients expose the Consul HTTP API via a static port (default 8500)
+bound to the host port. This enables all other pods on the node to connect
+to the node-local agent using the host IP that can be retrieved via the
+Kubernetes downward API.
```

Comment:
It would be awesome to give an example of how you do that here (or link to one if there is some other natural place for it) as even though I've read all the docs on this it's not _super_ clear how you actually configure tools like consul-template in this way.

I assume in another pod something like:

```
$ CONSUL_HTTP_ADDR="http://$HOST_IP:8500" consul kv get foo
```

But from docs it looks like you'd need to explicitly bind that ENV var to `status.hostIP` in the pod manifest right? Maybe something like:

```
spec:
  containers:
    - name: test-container
      ...
      env:
        - name: HOST_IP
          valueFrom:
            fieldRef:
              fieldPath: status.hostIP
```

Is there an easier way via the file-based Downward API volume? Seems that needs to be explicitly added to pod too.

I guess you considered this already but would it be possible to make an ingress controller (or whatever they are called now) that injected a CONSUL_HTTP_ADDR env variable into every pod in a templated way so it would pick up the host IP?

Reply:
I wonder how much work it would be to have all the major tools support talking to agent on Unix sockets and then just have a shared hostPath volume between the pods on a host (you'd have to add that volume to every pod that needs it but no different to downward API options it seems). That would also solve the "public listening" problem...

Just a thought for later.

Reply:
Good idea on both fronts. I'll add docs for how to access now.